
##IMPORTS
Imports = c("GenomicRanges","parallel")
lapply(Imports, library, character.only = T)

##FUNCTIONS

# make.list.GRanges.tiles ######################################################

#' Creates a list of tiles GRanges.
#' 
#' @param autosomes       An \code{integer} vector listing autosomes.
#' @param chr.len.table   A \code{data.frame} of 2 columns containing:
#'                        1 - autosomes as \code{integer}, 2 - autosomes sizes
#'                        in base pairs as \code{integer}. 
#' @param bins.size       The width of bins as an \code{integer} in base pairs.
#' @param ncores          An \code{integer} specifying the number of cores or
#'                        threads to speed up the process. By default equals 1.                      
#' @value a \code{list} of GRanges objects, 1 GRanges object by chromosome,
#'                        containing genome tiles ranges.
#' @author Yoann Pageaud.

make.list.GRanges.tiles<-function(autosomes, chr.len.table, bins.size,
                                  ncores = 1){
  List.chr.gr.tiles<-mclapply(autosomes, mc.cores = ncores, function(chr){
    makeGRangesFromDataFrame(data.frame(
      chr = rep(chr,length(seq(0,chr.len.table[chr,2]-bins.size,
                               by = bins.size))),
      start = seq(0, chr.len.table[chr,2]-bins.size,by = bins.size),
      end = seq(bins.size, chr.len.table[chr,2],by = bins.size)))
  })
  return(List.chr.gr.tiles)
}

# meth.tile.calc ###############################################################

#' Calculates the average methylation value of a genome tile.
#' 
#' @param overlaps.hits A \code{SortedByQueryHits} object generated by the
#'                      findOverlaps() function between the \code{GRanges} of
#'                      tiles and the \code{GRanges} of CpGs.
#' @param list.meth.df  A \code{list} of dataframes of CpGs, 1 dataframe by
#'                      chromosome. This object is generated by the core
#'                      function Read.Chroms.Meth().
#' @param autosome      An \code{integer} matching an autosome.
#' @param tile.number   An \code{integer} specifying the tile position in the
#'                      \code{GRanges} object containing all tiles.
#' @value the average methylation value of the tile as a \code{double}.
#' @author Yoann Pageaud.

meth.tile.calc<-function(overlaps.hits,list.meth.df,autosome,tile.number){
  #For each tile overlapped, create dataframe
  if(tile.number %in% queryHits(overlaps.hits)){
    df_m_local<-list.meth.df[[autosome]][
      subjectHits(overlaps.hits[queryHits(overlaps.hits)== tile.number]),c(3,4)]
    if(nrow(df_m_local) == 1){ #If only one row: calculate methylation
      meth_val<-df_m_local[, methylation:=V3/(V3+V4),]$methylation
    } else { #If multiple rows: calculate average methylation
      meth_val<-mean(df_m_local[, methylation:=V3/(V3+V4),]$methylation,na.rm=T)
    }
  } else { meth_val<-NA }
  return(meth_val)
}

# meth.tile.by.chr #############################################################

#' Lists average methylation of tiles by chromosomes
#' 
#' @param autosomes           An \code{integer} vector listing autosomes.
#' @param list.GRanges.tiles  A \code{list} containing the \code{GRanges}
#'                            objects of tiles, 1 \code{GRanges} object by
#'                            chromosome.
#' @param list.dframes.CpGs   A \code{list} of dataframes of CpGs, 1 dataframe
#'                            by chromosome. This object is generated by the
#'                            core function Read.Chroms.Meth().
#' @param ncores              An \code{integer} specifying the number of cores
#'                            or threads to speed up the loading. By default
#'                            equals 1.
#' 
#' @value A \code{list} of dataframes containing average methylation values of
#'                            tiles, with tiles by row.
#' @author Yoann Pageaud.

meth.tile.by.chr<-function(autosomes, list.GRanges.tiles, list.dframes.CpGs,
                           ncores=1){
  List_df_chr_tiles<-lapply(autosomes, function(chr){
    cat(paste0("\tchr",chr,"\n"))
    #Load Granges of tiles for the chromosome
    gr_tiles<-list.GRanges.tiles[[chr]]
    #Create Granges of CpGs for chromosome
    gr_MC<-makeGRangesFromDataFrame(data.frame(
      chr = rep(chr, nrow(list.dframes.CpGs[[chr]])),
      start = list.dframes.CpGs[[chr]][["V1"]]+1,
      end = list.dframes.CpGs[[chr]][["V1"]]+1))
    #Get Overlaps between tiles and CpGs
    hits<-findOverlaps(gr_tiles,gr_MC)
    #Calculate methylation value of the tiles
    meth_val<-unlist(mclapply(seq_along(gr_tiles),mc.cores = ncores,function(i){
      meth.tile.calc(overlaps.hits = hits, list.meth.df = list.dframes.CpGs,
                     autosome = chr, tile.number = i)
    }))
    #Get tiles ranges and their methylation in dataframe
    df_tiles<-as.data.frame(ranges(gr_tiles))
    return(cbind(df_tiles[,c(1,2)],meth_val))
  })
  #Map cbind chromosomes to list of dataframes
  List_df_chr_tiles<-Map(cbind,chr = autosomes,List_df_chr_tiles)
  return(List_df_chr_tiles)
}
